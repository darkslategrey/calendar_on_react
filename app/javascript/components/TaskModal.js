import 'react-datepicker/dist/react-datepicker.css';

import React from 'react';
import Modal from 'react-modal';
import DatePicker, { registerLocale } from 'react-datepicker';

import en from 'date-fns/locale/en-GB';
import { connect } from "react-redux";

registerLocale('en', en);

Modal.setAppElement('body')

class TaskModal extends React.Component {
  constructor (props) {
    super(props);

    this.datePickerDateFormat = 'dd/M/yyyy';
  }

  handleSubmit(event) {
    const id = this.props.formData.id;
    const method = id ? 'PUT' : 'POST';

    fetch(`/tasks/${id ? id : ''}`, {
      method: method,
      headers:
        {
         'Accept': 'application/json',
         'Content-Type': 'application/json',
        },
      body: JSON.stringify(this.props.formData)})
    .then(response =>  {
       if (!response.ok)
            throw Error(response.statusText);

        alert(`Task ${id ? 'updated' : 'added'} successfully!`)
        location.reload()
     })
    .catch((error) => {
      alert('Error')
    });

    event.preventDefault();
  }

  render () {
    return (
      <div className='calendar__task-modal'>
        <button
          className='calendar__add-task-btn'
          onClick={this.props.handleOpenModal}
        >
          Add Task
        </button>
        <Modal
           isOpen={this.props.showModal}
           contentLabel='Minimal Modal Example'
        >
        <form
          onSubmit={this.handleSubmit.bind(this)}
          className='calendar__modal-form'
        >
          <div className='calendar__modal-form-group'>
            <label>Name:</label>
            <input
              type='text'
              id='calendar__modal-form-name'
              onChange={(e) => this.props.handleFieldChange(e, 'name')}
              value={this.props.formData.name}
            />
          </div>
          <div className='calendar__modal-form-group'>
            <label>Start Date:</label>
            <DatePicker
              locale='en'
              selectsStart
              startDate={this.props.datePickerData.startDate}
              dateFormat={this.datePickerDateFormat}
              selected={this.props.datePickerData.startDate}
              onChange={(d) => this.props.handleDateChange(d, 'startDate')}
            />
          </div>
          <div className='calendar__modal-form-group'>
            <label>End Date:</label>
            <DatePicker
              locale='en'
              selectsEnd
              endDate={this.props.datePickerData.endDate}
              minDate={this.props.datePickerData.startDate}
              dateFormat={this.datePickerDateFormat}
              selected={this.props.datePickerData.endDate}
              onChange={(d) => this.props.handleDateChange(d, 'endDate')}
            />
          </div>
          <input type='submit' id='calendar__modal-submit' value='Submit'/>
          <button onClick={this.props.handleCloseModal}>Close Modal</button>
        </form>
        </Modal>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    showModal: state.taskModal.showModal,
    formData: state.taskModal.formData,
    datePickerData: state.taskModal.datePickerData
  };
}

const mapDispatchToProps = dispatch => {
  return {
    handleOpenModal: () => dispatch(
      {
        type: 'TASK_MODAL_OPEN_NEW',
        payload:
          {
            showModal: true
          }
      }
    ),
    handleCloseModal: () => dispatch(
      {
        type: 'TASK_MODAL_CLOSE',
        payload:
          {
            showModal: false
          }
      }
    ),
    handleFieldChange: (event, key) => dispatch(
      {
        type: 'TASK_MODAL_FIELD_CHANGE',
        payload:
          {
            key: key,
            value: event.target.value
          }
      }
    ),
    handleDateChange: (date, key) => dispatch(
      {
        type: 'TASK_MODAL_DATE_CHANGE',
        payload:
          {
            key: key,
            date: date
          }
      }
    )
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(TaskModal);
