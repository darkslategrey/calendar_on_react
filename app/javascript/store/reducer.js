import SharedMethodsService from '../services/sharedMethodsService';

const initialState = {
  taskModal: {
    showModal: false,
    formData: {
      task_id: null,
      name: '',
      start_date: null,
      end_date: null
    },
    datePickerData:
      {
        startDate: null,
        endDate: null}
      },
    taskElements: {},
    year: 2019,
    month: 11,
    hoveredTaskId: null,
    unhoveredTaskId: null,
    tasksJson: [],
    weekHeights: {},
    dateElementsRefs: {}
  }

const reducer = (state = initialState, action) => {
  let tmp = null;

  switch(action.type) {
    case 'ADD_DATE_ELEMENTS_REFS':
      return  Object.assign({}, state, action.payload);
      break;
    case 'ADD_TASK_ELEMENTS':
      return  Object.assign({}, state, action.payload);
      break;
    case 'SET_TASK_JSON_DATA':
      return Object.assign(
        {},
        state,
        action.payload,
        {
          taskElements: {},
          weekHeights: {},
          hoveredTaskId: null,
          unhoveredTaskId: null
        }
      );
      break;
    case 'TASK_MODAL_CLOSE':
      return {...state, taskModal: initialState.taskModal};
      break;
    case 'TASK_MODAL_DATE_CHANGE':
      const formDataDateChange = {...state.taskModal.formData};
      const datePickerData = {...state.taskModal.datePickerData};
      const year = action.payload.date.getFullYear();
      const month = action.payload.date.getMonth() + 1;
      const date = action.payload.date.getDate();
      const dateString = `${date}-${month}-${year}`;

      formDataDateChange[action.payload.key.replace('Date', '_date')] = dateString;
      datePickerData[action.payload.key] = action.payload.date;

      return {
        ...state,
        taskModal:
          {
            ...state.taskModal,
            datePickerData: datePickerData,
            formData: formDataDateChange
          }
      };
      break;
    case 'TASK_MODAL_FIELD_CHANGE':
      const formDataFieldChange = {...state.taskModal.formData};
      formDataFieldChange[action.payload.key] = action.payload.value;

      return {
        ...state,
        taskModal:
          {
            ...state.taskModal,
            formData: formDataFieldChange
          }
      };
      break;
    case 'TOGGLE_HOVER_EFFECT':
      return  Object.assign({}, state, action.payload);
      break;
    case 'TASK_MODAL_OPEN_EDIT':
      const startDateObj = new Date(
        ...SharedMethodsService.getDateValues(action.payload.task.startDate)
      );
      const endDateObj = new Date(
        ...SharedMethodsService.getDateValues(action.payload.task.endDate)
      );

      return  {
        ...state,
        taskModal:
          Object.assign(
            {},
            state.taskModal,
            { showModal: true },
            { datePickerData:
              { startDate: startDateObj,
                endDate: endDateObj
              }
            },
            { formData:
              { id: action.payload.task.id,
                name: action.payload.task.name || '',
                start_date: action.payload.task.startDate,
                end_date: action.payload.task.endDate
              }
            }
          )
        };
      break;
    case 'TASK_MODAL_OPEN_NEW':
      return  {...state, taskModal: {...state.taskModal, showModal: true}};
      break;
    default:
      return state;
  }
};

export {reducer as default, initialState as initialState};
