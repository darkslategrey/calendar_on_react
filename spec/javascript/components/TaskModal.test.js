import React from 'react';
import TaskModal from '../../../app/javascript/components/TaskModal';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store'
import { mount } from 'enzyme';
import { initialState } from '../../../app/javascript/store/reducer';

describe('<TaskModal />', () => {
  it('renders', () => {
    const mockStore = configureStore()
    const store = mockStore({taskModal: {...initialState.taskModal, showModal: true}});
    const wrapper = mount(<Provider store={store}><TaskModal /></Provider>);

    expect(wrapper).toMatchSnapshot();
  });
})
