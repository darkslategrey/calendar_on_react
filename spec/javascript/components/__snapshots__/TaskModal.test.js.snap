// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskModal /> renders 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(TaskModal)>
    <TaskModal
      datePickerData={
        Object {
          "endDate": null,
          "startDate": null,
        }
      }
      formData={
        Object {
          "end_date": null,
          "name": "",
          "start_date": null,
          "task_id": null,
        }
      }
      handleCloseModal={[Function]}
      handleDateChange={[Function]}
      handleFieldChange={[Function]}
      handleOpenModal={[Function]}
      showModal={true}
    >
      <div
        className="calendar__task-modal"
      >
        <button
          className="calendar__add-task-btn"
          onClick={[Function]}
        >
          Add Task
        </button>
        <Modal
          ariaHideApp={true}
          bodyOpenClassName="ReactModal__Body--open"
          closeTimeoutMS={0}
          contentLabel="Minimal Modal Example"
          isOpen={true}
          parentSelector={[Function]}
          portalClassName="ReactModalPortal"
          role="dialog"
          shouldCloseOnEsc={true}
          shouldCloseOnOverlayClick={true}
          shouldFocusAfterRender={true}
          shouldReturnFocusAfterClose={true}
        >
          <Portal
            containerInfo={
              <div
                class="ReactModalPortal"
              >
                <div
                  class="ReactModal__Overlay ReactModal__Overlay--after-open"
                  style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(255, 255, 255, 0.75);"
                >
                  <div
                    aria-label="Minimal Modal Example"
                    class="ReactModal__Content ReactModal__Content--after-open"
                    role="dialog"
                    style="position: absolute; top: 40px; left: 40px; right: 40px; bottom: 40px; border: 1px solid #ccc; background: rgb(255, 255, 255); overflow: auto; border-radius: 4px; outline: none; padding: 20px;"
                    tabindex="-1"
                  >
                    <form
                      class="calendar__modal-form"
                    >
                      <div
                        class="calendar__modal-form-group"
                      >
                        <label>
                          Name:
                        </label>
                        <input
                          id="calendar__modal-form-name"
                          type="text"
                          value=""
                        />
                      </div>
                      <div
                        class="calendar__modal-form-group"
                      >
                        <label>
                          Start Date:
                        </label>
                        <div
                          class="react-datepicker-wrapper"
                        >
                          <div
                            class="react-datepicker__input-container"
                          >
                            <input
                              class=""
                              type="text"
                              value=""
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class="calendar__modal-form-group"
                      >
                        <label>
                          End Date:
                        </label>
                        <div
                          class="react-datepicker-wrapper"
                        >
                          <div
                            class="react-datepicker__input-container"
                          >
                            <input
                              class=""
                              type="text"
                              value=""
                            />
                          </div>
                        </div>
                      </div>
                      <input
                        id="calendar__modal-submit"
                        type="submit"
                        value="Submit"
                      />
                      <button>
                        Close Modal
                      </button>
                    </form>
                  </div>
                </div>
              </div>
            }
          >
            <ModalPortal
              ariaHideApp={true}
              bodyOpenClassName="ReactModal__Body--open"
              closeTimeoutMS={0}
              contentLabel="Minimal Modal Example"
              defaultStyles={
                Object {
                  "content": Object {
                    "WebkitOverflowScrolling": "touch",
                    "background": "#fff",
                    "border": "1px solid #ccc",
                    "borderRadius": "4px",
                    "bottom": "40px",
                    "left": "40px",
                    "outline": "none",
                    "overflow": "auto",
                    "padding": "20px",
                    "position": "absolute",
                    "right": "40px",
                    "top": "40px",
                  },
                  "overlay": Object {
                    "backgroundColor": "rgba(255, 255, 255, 0.75)",
                    "bottom": 0,
                    "left": 0,
                    "position": "fixed",
                    "right": 0,
                    "top": 0,
                  },
                }
              }
              isOpen={true}
              parentSelector={[Function]}
              portalClassName="ReactModalPortal"
              role="dialog"
              shouldCloseOnEsc={true}
              shouldCloseOnOverlayClick={true}
              shouldFocusAfterRender={true}
              shouldReturnFocusAfterClose={true}
              style={
                Object {
                  "content": Object {},
                  "overlay": Object {},
                }
              }
            >
              <div
                className="ReactModal__Overlay ReactModal__Overlay--after-open"
                onClick={[Function]}
                onMouseDown={[Function]}
                style={
                  Object {
                    "backgroundColor": "rgba(255, 255, 255, 0.75)",
                    "bottom": 0,
                    "left": 0,
                    "position": "fixed",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <div
                  aria-label="Minimal Modal Example"
                  className="ReactModal__Content ReactModal__Content--after-open"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                  role="dialog"
                  style={
                    Object {
                      "WebkitOverflowScrolling": "touch",
                      "background": "#fff",
                      "border": "1px solid #ccc",
                      "borderRadius": "4px",
                      "bottom": "40px",
                      "left": "40px",
                      "outline": "none",
                      "overflow": "auto",
                      "padding": "20px",
                      "position": "absolute",
                      "right": "40px",
                      "top": "40px",
                    }
                  }
                  tabIndex="-1"
                >
                  <form
                    className="calendar__modal-form"
                    onSubmit={[Function]}
                  >
                    <div
                      className="calendar__modal-form-group"
                    >
                      <label>
                        Name:
                      </label>
                      <input
                        id="calendar__modal-form-name"
                        onChange={[Function]}
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      className="calendar__modal-form-group"
                    >
                      <label>
                        Start Date:
                      </label>
                      <t
                        allowSameDay={false}
                        dateFormat="dd/M/yyyy"
                        dateFormatCalendar="LLLL yyyy"
                        disabled={false}
                        disabledKeyboardNavigation={false}
                        dropdownMode="scroll"
                        inlineFocusSelectedMonth={false}
                        locale="en"
                        monthsShown={1}
                        nextMonthButtonLabel="Next Month"
                        nextYearButtonLabel="Next Year"
                        onBlur={[Function]}
                        onCalendarClose={[Function]}
                        onCalendarOpen={[Function]}
                        onChange={[Function]}
                        onClickOutside={[Function]}
                        onFocus={[Function]}
                        onInputClick={[Function]}
                        onInputError={[Function]}
                        onKeyDown={[Function]}
                        onMonthChange={[Function]}
                        onSelect={[Function]}
                        onYearChange={[Function]}
                        preventOpenOnFocus={false}
                        previousMonthButtonLabel="Previous Month"
                        previousYearButtonLabel="Previous Year"
                        readOnly={false}
                        renderDayContents={[Function]}
                        selected={null}
                        selectsStart={true}
                        shouldCloseOnSelect={true}
                        showMonthYearPicker={false}
                        showPopperArrow={true}
                        showPreviousMonths={false}
                        showQuarterYearPicker={false}
                        showTimeInput={false}
                        showTimeSelect={false}
                        startDate={null}
                        strictParsing={false}
                        timeCaption="Time"
                        timeInputLabel="Time"
                        timeIntervals={30}
                        withPortal={false}
                      >
                        <t
                          hidePopper={true}
                          popperComponent={null}
                          popperModifiers={
                            Object {
                              "preventOverflow": Object {
                                "boundariesElement": "viewport",
                                "enabled": true,
                                "escapeWithReference": true,
                              },
                            }
                          }
                          popperPlacement="bottom-start"
                          popperProps={Object {}}
                          targetComponent={
                            <div
                              className="react-datepicker__input-container"
                            >
                              <input
                                className=""
                                disabled={false}
                                onBlur={[Function]}
                                onChange={[Function]}
                                onClick={[Function]}
                                onFocus={[Function]}
                                onKeyDown={[Function]}
                                readOnly={false}
                                type="text"
                                value=""
                              />
                            </div>
                          }
                        >
                          <Manager
                            className="react-datepicker-manager"
                          >
                            <Reference>
                              <InnerReference
                                setReferenceNode={[Function]}
                              >
                                <div
                                  className="react-datepicker-wrapper"
                                >
                                  <div
                                    className="react-datepicker__input-container"
                                  >
                                    <input
                                      className=""
                                      disabled={false}
                                      onBlur={[Function]}
                                      onChange={[Function]}
                                      onClick={[Function]}
                                      onFocus={[Function]}
                                      onKeyDown={[Function]}
                                      readOnly={false}
                                      type="text"
                                      value=""
                                    />
                                  </div>
                                </div>
                              </InnerReference>
                            </Reference>
                          </Manager>
                        </t>
                      </t>
                    </div>
                    <div
                      className="calendar__modal-form-group"
                    >
                      <label>
                        End Date:
                      </label>
                      <t
                        allowSameDay={false}
                        dateFormat="dd/M/yyyy"
                        dateFormatCalendar="LLLL yyyy"
                        disabled={false}
                        disabledKeyboardNavigation={false}
                        dropdownMode="scroll"
                        endDate={null}
                        inlineFocusSelectedMonth={false}
                        locale="en"
                        minDate={null}
                        monthsShown={1}
                        nextMonthButtonLabel="Next Month"
                        nextYearButtonLabel="Next Year"
                        onBlur={[Function]}
                        onCalendarClose={[Function]}
                        onCalendarOpen={[Function]}
                        onChange={[Function]}
                        onClickOutside={[Function]}
                        onFocus={[Function]}
                        onInputClick={[Function]}
                        onInputError={[Function]}
                        onKeyDown={[Function]}
                        onMonthChange={[Function]}
                        onSelect={[Function]}
                        onYearChange={[Function]}
                        preventOpenOnFocus={false}
                        previousMonthButtonLabel="Previous Month"
                        previousYearButtonLabel="Previous Year"
                        readOnly={false}
                        renderDayContents={[Function]}
                        selected={null}
                        selectsEnd={true}
                        shouldCloseOnSelect={true}
                        showMonthYearPicker={false}
                        showPopperArrow={true}
                        showPreviousMonths={false}
                        showQuarterYearPicker={false}
                        showTimeInput={false}
                        showTimeSelect={false}
                        strictParsing={false}
                        timeCaption="Time"
                        timeInputLabel="Time"
                        timeIntervals={30}
                        withPortal={false}
                      >
                        <t
                          hidePopper={true}
                          popperComponent={null}
                          popperModifiers={
                            Object {
                              "preventOverflow": Object {
                                "boundariesElement": "viewport",
                                "enabled": true,
                                "escapeWithReference": true,
                              },
                            }
                          }
                          popperPlacement="bottom-start"
                          popperProps={Object {}}
                          targetComponent={
                            <div
                              className="react-datepicker__input-container"
                            >
                              <input
                                className=""
                                disabled={false}
                                onBlur={[Function]}
                                onChange={[Function]}
                                onClick={[Function]}
                                onFocus={[Function]}
                                onKeyDown={[Function]}
                                readOnly={false}
                                type="text"
                                value=""
                              />
                            </div>
                          }
                        >
                          <Manager
                            className="react-datepicker-manager"
                          >
                            <Reference>
                              <InnerReference
                                setReferenceNode={[Function]}
                              >
                                <div
                                  className="react-datepicker-wrapper"
                                >
                                  <div
                                    className="react-datepicker__input-container"
                                  >
                                    <input
                                      className=""
                                      disabled={false}
                                      onBlur={[Function]}
                                      onChange={[Function]}
                                      onClick={[Function]}
                                      onFocus={[Function]}
                                      onKeyDown={[Function]}
                                      readOnly={false}
                                      type="text"
                                      value=""
                                    />
                                  </div>
                                </div>
                              </InnerReference>
                            </Reference>
                          </Manager>
                        </t>
                      </t>
                    </div>
                    <input
                      id="calendar__modal-submit"
                      type="submit"
                      value="Submit"
                    />
                    <button
                      onClick={[Function]}
                    >
                      Close Modal
                    </button>
                  </form>
                </div>
              </div>
            </ModalPortal>
          </Portal>
        </Modal>
      </div>
    </TaskModal>
  </Connect(TaskModal)>
</Provider>
`;
